FIFO attempt

Rinnerdos opened the pull request:
This attempt doesn't implement FIFO yet, but I hope that I created the first steps to be able to do it. To me it seems that now only the most perished products need to be passed to the filter, which should be relatively easy with the duplicates objects which gives the duplicates in a container, and their index. I might be wrong of course, and it also might be slightly redundant/too complex. Also, the items can be spread over multiple containers, and I'm not sure if that is easy to incorprate in my method.

Rinnerdos committed:
FIFO attempt, not working yet

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
+        // Creates an object with the duplicate name,
+        // and their index in the original array.
+
+        Array.prototype.getDuplicates = function () {
```
Not really problematic, but don't do this here. You're extending the Array prototype in a place where this is _not_ expected. Better to do this in main.js, by including a new file there called e.g. `extensions.js`

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
+
+        Array.prototype.getDuplicates = function () {
+            let duplicates = {};
+            for (var i = 0; i < this.length; i++) {
```
`let`.

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
+        let perishTimeArray = [];
+        let perishNameArray = [];
+        let perishDuplicates = {};
+        for (let i = 0; i < this.items.length; i++) {
```
`this.items.forEach((product) => {...})`?

N-Wouda commented:
Attempt to align/structure your code a bit better. Not just in terms of _what_ you wrote, but also _how_ it looks. This is a bit messy.

N-Wouda committed:
OrderProcessController

Rinnerdos committed:
Start of structure

Rinnerdos committed:
Almost..

I think I'm nearly there for FIFO. Yes, I know the form is pretty rough, but I just wanted it to work first. I can find the index of the container which has the most perished prodoct of interest, but I'm not sure how to make sure the product is removed from that container.  What I now have in the last line with the removing is not working (obviously), it just leads to the normal removal system.

Rinnerdos committed:
Attempt to neat things up

Now product get removed from most perished container, though it leads to overload when the most perished container can only partially fulfill order

N-Wouda committed:
Return value

N-Wouda commented:
How is this going?

Rinnerdos committed:
Perished Quantity

I now have the quantity of the item which is most perished, only unsure how to pass this to the removeItem function. Do I need to change something in the container.js?

Rinnerdos committed:
Don't think it was this

Rinnerdos commented:
See commits

Rinnerdos committed:
Try in container.js

So I tried changing some things in container.js, to no avail. How does the removeItem work exactly? I did some console logging and I believe the item.values.quanity will give the quantity of the first object in the array of items, which is the same as removedQuantity, making my code redundant. But what happens when the function returns?

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
 
-            const removedQuantity = Math.min(product.values.quantity, item.values.quantity);
+            if (product.values.isPerishable === true) {
+                const removedQuantity = Math.min(product.values.quantity, perishedQuantity)
```
Product is the external value, perishedQuantity is.. too? You need to check against `item`, which is the product actually in the container.

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
             }
 
-            const removedQuantity = Math.min(product.values.quantity, item.values.quantity);
+            if (product.values.isPerishable === true) {
```
```javascript
if (product.values.isPerishable) {
```

Rinnerdos replied:
Ah yeah, forgot about that :D

Rinnerdos committed:
perishedProduct

I now am passing the 'perishedProduct' to the removeItem function, though not with much success. When there are 2 or more items of the same name in a container, it gets stuck in an infinite loop. Console logging product.values.quantity during this reveals this value will jump between the initial positive 9 (from the let's say 9 bananas in the order), to -9, to 9, -9, etc. 

I'm obviously doing something wrong, though I cant find what. Am I on the right track?

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -36,16 +36,16 @@ export default class Container extends StorageCore
-                const removedQuantity = Math.min(product.values.quantity, perishedQuantity)
-             
+            if (product.values.isPerishable) {
+                const removedQuantity = perishedProduct.values.quantity;
```
`Math.min(product.values.quantity, item.values.quantity, perishedQuantity);`

You can only delete the bare minimum of this (but you should not use `perishQuantity`, but rather update the value on `product` outside this function).

Rinnerdos replied:
Yeah that was dumb of me. I now made it 

```Math.min(product.values.quantity, perishedProduct.values.quantity);```

But I just keep struggling with that it doesn't stop removing from a container after the first object has been removed (when I want it to).

N-Wouda commented on src/models/classes/container.js:
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
+                item.values.quantity -= removedQuantity;
 
-            return item.values.quantity;
+                return item.values.quantity;
```
You should return `product.values.quantity`

Don't return here, use the one at the end of the function.

Rinnerdos committed:
Not a functional update

But with this I hope to explain what's going wrong, during the meeting. Basically it's staying in the removeItem function until product.values.quantity = 0, and I don't know why.

Rinnerdos committed:
FIFO working

It works!
